ICU-20602 add LocaleBuilder::copyErrorTo

FrankYFTang opened the pull request:
<!--
Thank you for your pull request.
Please see http://site.icu-project.org/processes/contribute for general
information on contributing to ICU.

You will be automatically asked to sign the contributors license before the PR is accepted.
- sign: https://cla-assistant.io/unicode-org/icu
- license: http://www.unicode.org/copyright.html#License
-->

##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20602
- [X] Updated PR title and link in previous line to include Issue number
- [X] Issue accepted
- [X] Tests included
- [X] Documentation is changed or added

sffc commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -433,4 +433,9 @@ Locale LocaleBuilder::build(UErrorCode& errorCode)
     return product;
 }
 
+UBool LocaleBuilder::copyErrorTo(UErrorCode &outErrorCode) {
```
This does not

> Preserves older error codes in the outErrorCode.

Please add

```cpp

        if (U_FAILURE(outErrorCode)) {
            // Do not overwrite the older error code
            return TRUE;
        }
```

FrankYFTang replied:
DONE

sffc commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -433,4 +433,9 @@ Locale LocaleBuilder::build(UErrorCode& errorCode)
 
+UBool LocaleBuilder::copyErrorTo(UErrorCode &outErrorCode) {
+    outErrorCode = status_;
+    return U_FAILURE(outErrorCode) ? TRUE : FALSE;
```
U_FAILURE already returns a UBool.  No need to pass it through another ternary.

FrankYFTang replied:
DONE

sffc commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -190,6 +197,9 @@ void LocaleBuilderTest::TestLocaleBuilder() {
         while (true) {
+            status = U_ZERO_ERROR;
+            UErrorCode copyStatus = U_ZERO_ERROR;
+            bld.copyErrorTo(copyStatus);
```
You copy the error before you start mutating the LocaleBuilder.  Wouldn't you want to put this after you call all the various builder methods?

I would also like to see a concrete test where the error code is not just U_ZERO_ERROR.  Give it bad input on purpose, and check that the error code is as expected.  Maybe you already have that, but it's not obvious from looking at the test.

sffc commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -55,10 +55,17 @@ void LocaleBuilderTest::runIndexedTest( int32_t index, UBool exec, const char* &
 
 void LocaleBuilderTest::Verify(LocaleBuilder& bld, const char* expected, const char* msg) {
     UErrorCode status = U_ZERO_ERROR;
```
FYI for the future: in tests, you can and should use IcuTestErrorCode from testlog.h.

FrankYFTang commented:
PTAL

FrankYFTang commented:
> I don't see:
> 
> > I would also like to see a concrete test where the error code is not just U_ZERO_ERROR. Give it bad input on purpose, and check that the error code is as expected. Maybe you already have that, but it's not obvious from looking at the test.

Inside the LocaleBuilderTest::Verify function

sffc commented:
Ok.  Trying to read your verify function.  So you have data with "X", and then you have

```cpp
            else if (strcmp("X", method) == 0) {
                if (U_SUCCESS(status)) {
                    errln("FAIL: No error return - test case: %s", actions.c_str());
                }
```

And then you also have

```cpp
            if (status != copyStatus) {
                errln("copyErrorTo not matching");
            }
```

However, between those two lines, you have

```cpp
            if (U_FAILURE(status)) {
                if (strcmp("X", testCase[i]) == 0) {
                    // This failure is expected
                    break;
```

Can you verify that `copyStatus == status` also in the failure case?

FrankYFTang commented:
> Ok. Trying to read your verify function. So you have data with "X", and then you have
> 
> ```c++
>             else if (strcmp("X", method) == 0) {
>                 if (U_SUCCESS(status)) {
>                     errln("FAIL: No error return - test case: %s", actions.c_str());
>                 }
> ```
> 
> And then you also have
> 
> ```c++
>             if (status != copyStatus) {
>                 errln("copyErrorTo not matching");
>             }
> ```
> 
> However, between those two lines, you have
> 
> ```c++
>             if (U_FAILURE(status)) {
>                 if (strcmp("X", testCase[i]) == 0) {
>                     // This failure is expected
>                     break;
> ```
> 
> Can you verify that `copyStatus == status` also in the failure case?

DONE. PTAL

FrankYFTang commented:
> It looks like you used the new squash commit button. There was a bug in that button which I fixed yesterday at 9pm PDT, but it looks like you clicked the button before I fixed it.
> 
> Please force-push from your local branch (reset this PR to what it was before you squashed), then you can try the squash flow again.

PTAL

sffc commented on icu4c/source/common/unicode/localebuilder.h:
```diff
@@ -276,6 +276,17 @@ class U_COMMON_API LocaleBuilder : public UObject {
+     * Sets the UErrorCode if an error occurred while recording sets.
+     * Preserves older error codes in the outErrorCode.
+     * @param outErrorCode Set to an error code if it does not contain one already
+     *                  and an error occurred while recording sets.
```
Suggested documentation of `@param outErrorCode`:

```
Set to an error code that occurred while setting subtags.
Unchanged if there is no such error or if outErrorCode
already contained an error.
```

FrankYFTang replied:
DONE PTAL
